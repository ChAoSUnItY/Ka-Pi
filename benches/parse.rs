use criterion::{black_box, criterion_group, criterion_main, Criterion};
use ka_pi::parse::to_class;

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("parse main", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Main.class"
    )))));

    c.bench_function("parse enum", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Enum.class"
    )))));

    c.bench_function("parse record", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Record.class"
    )))));

    c.bench_function("parse visible annotation", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/VisibleAnnotation.class"
    )))));

    c.bench_function("parse invisible annotation", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/InvisibleAnnotation.class"
    )))));

    c.bench_function("parse annotation target", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/AnnotationTarget.class"
    )))));
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
