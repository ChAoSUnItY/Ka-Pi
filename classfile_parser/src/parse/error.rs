use std::io;

use thiserror::Error;

use crate::node::opcode::Opcode;

pub type ParseResult<T> = Result<T, ParseError>;

#[derive(Debug, Error)]
pub enum ParseError {
    #[error(transparent)]
    Io(#[from] io::Error),
    #[error("expected [0xCA, 0xFE, 0xBA, 0xBE] at the header of class file but got {0:?}")]
    MismatchedMagicNumber([u8; 4]),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBoundUsize(&'static str, Vec<&'static str>, usize),
    #[error("attempt to match {0} in {1:?} but got '{2}'")]
    MatchOutOfBoundChar(&'static str, Vec<char>, char),
    #[error("attempt to match opcode but got opcode value {0}")]
    MatchOutOfBoundOpcode(u8),
    #[error("attempt to match sub opcode in wide opcode but got {0}, while {1:?} is allowed as sub opcode of wide opcode")]
    MatchOutOfBoundWideOpcode(u8, Vec<Opcode>),
    #[error("the parse for segment is finished but still remains {0} bytes")]
    Remains(usize),
    #[error("expected {1:?} but got '{0}'")]
    MismatchedCharacter(char, Vec<char>),
    #[error("attempt to parse {0} but is incomplete")]
    OutOfBound(&'static str),
}
