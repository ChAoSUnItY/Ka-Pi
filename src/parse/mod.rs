use std::io::Read;

pub use error::{ParseError, ParseResult};
pub use signature::{class_signature, field_signature, method_signature};

use crate::node::class::Class;

pub(crate) mod attribute;
pub(crate) mod class;
pub(crate) mod constant;
pub(crate) mod error;
pub(crate) mod field;
pub(crate) mod method;
pub(crate) mod signature;

/// Parses class file based on given options.
pub fn to_class<R: Read>(class_bytes: &mut R, option: ParsingOption) -> ParseResult<Class> {
    let class = class::class(class_bytes, option)?;

    let mut remain = vec![];

    class_bytes.read_to_end(&mut remain)?;

    if !remain.is_empty() {
        Err(ParseError::Remains(remain.len()))
    } else {
        Ok(class)
    }
}

/// Parsing options allows marking some parsing phase optional.
#[derive(Debug, Default)]
pub struct ParsingOption {
    parse_attribute: bool,
}

impl ParsingOption {
    /// Skips on attribute struct parsing.
    pub fn parse_attribute(mut self) -> Self {
        self.parse_attribute = true;

        self
    }
}
