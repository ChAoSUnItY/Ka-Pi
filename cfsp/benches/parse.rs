use cfsp::parse::{to_class, ParsingOption};
use std::fs;
use std::io::Cursor;

use criterion::{criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion, Throughput};

fn parsing_classes(c: &mut Criterion) {
    let mut g = c.benchmark_group("parse class");

    let compiled_source_folder =
        fs::read_dir("../compiled_source/out/production/compiled_source").unwrap();

    for entry in compiled_source_folder {
        let entry = entry.as_ref().unwrap().path();
        let class_name = entry.file_name().unwrap().to_str().unwrap();
        let class_bytes = fs::read(&entry).unwrap();

        g.throughput(Throughput::Bytes(class_bytes.len() as u64));
        g.bench_with_input(
            BenchmarkId::from_parameter(class_name),
            &class_bytes,
            |b, bytes| {
                b.iter_batched(
                    || Cursor::new(bytes),
                    |mut cursor| {
                        to_class(&mut cursor, ParsingOption::default().parse_attribute())
                            .expect("Parsing fails on benchmarking");
                    },
                    BatchSize::SmallInput,
                );
            },
        );

        g.bench_with_input(
            BenchmarkId::from_parameter(format!("{class_name} (No attribute parsing)")),
            &class_bytes,
            |b, bytes| {
                b.iter_batched(
                    || Cursor::new(bytes),
                    |mut cursor| {
                        to_class(&mut cursor, ParsingOption::default())
                            .expect("Parsing fails on benchmarking");
                    },
                    BatchSize::SmallInput,
                );
            },
        );
    }
}

criterion_group!(benches, parsing_classes);
criterion_main!(benches);
