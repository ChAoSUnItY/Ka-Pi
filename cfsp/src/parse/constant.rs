use std::cell::RefCell;
use std::io::Read;

use byteorder::{BigEndian, ReadBytesExt};

use crate::node::constant;
use crate::node::constant::{
    Class, Constant, ConstantPool, Double, Dynamic, FieldRef, Float, Integer, InterfaceMethodRef,
    InvokeDynamic, Long, MethodHandle, MethodRef, MethodType, Module, NameAndType, Package, Utf8,
};
use crate::parse::error::{ParseError, ParseResult};

pub(super) fn constant_pool<R: Read>(input: &mut R) -> ParseResult<(u16, ConstantPool)> {
    let len = input.read_u16::<BigEndian>()?;
    let mut constant_pool = ConstantPool::with_capacity(len);
    let mut constant_counter = 0;

    while constant_counter < len - 1 {
        let constant = constant(input)?;

        if constant.occupies_2_slots() {
            constant_counter += 2;
        } else {
            constant_counter += 1;
        }

        constant_pool.add(constant);
    }

    Ok((len, constant_pool))
}

fn constant<R: Read>(input: &mut R) -> ParseResult<Constant> {
    let tag = input.read_u8()?;

    let constant = match tag {
        1 => {
            let length = input.read_u16::<BigEndian>()?;
            let mut bytes = vec![0; length as usize];

            input.read_exact(&mut bytes)?;

            Constant::Utf8(Utf8 {
                length,
                bytes,
                string: RefCell::new(None),
            })
        }
        3 => {
            let mut bytes = [0; 4];

            input.read_exact(&mut bytes)?;

            Constant::Integer(Integer { bytes })
        }
        4 => {
            let mut bytes = [0; 4];

            input.read_exact(&mut bytes)?;

            Constant::Float(Float { bytes })
        }
        5 => {
            let mut high_bytes = [0; 4];
            let mut low_bytes = [0; 4];

            input.read_exact(&mut high_bytes)?;
            input.read_exact(&mut low_bytes)?;

            Constant::Long(Long {
                high_bytes,
                low_bytes,
            })
        }
        6 => {
            let mut high_bytes = [0; 4];
            let mut low_bytes = [0; 4];

            input.read_exact(&mut high_bytes)?;
            input.read_exact(&mut low_bytes)?;

            Constant::Double(Double {
                high_bytes,
                low_bytes,
            })
        }
        7 => {
            let name_index = input.read_u16::<BigEndian>()?;

            Constant::Class(Class { name_index })
        }
        8 => {
            let string_index = input.read_u16::<BigEndian>()?;

            Constant::String(constant::String { string_index })
        }
        9 => {
            let class_index = input.read_u16::<BigEndian>()?;
            let name_and_type_index = input.read_u16::<BigEndian>()?;

            Constant::FieldRef(FieldRef {
                class_index,
                name_and_type_index,
            })
        }
        10 => {
            let class_index = input.read_u16::<BigEndian>()?;
            let name_and_type_index = input.read_u16::<BigEndian>()?;

            Constant::MethodRef(MethodRef {
                class_index,
                name_and_type_index,
            })
        }
        11 => {
            let class_index = input.read_u16::<BigEndian>()?;
            let name_and_type_index = input.read_u16::<BigEndian>()?;

            Constant::InterfaceMethodRef(InterfaceMethodRef {
                class_index,
                name_and_type_index,
            })
        }
        12 => {
            let name_index = input.read_u16::<BigEndian>()?;
            let type_index = input.read_u16::<BigEndian>()?;

            Constant::NameAndType(NameAndType {
                name_index,
                type_index,
            })
        }
        15 => {
            let reference_kind = input.read_u8()?;
            let reference_index = input.read_u16::<BigEndian>()?;

            Constant::MethodHandle(MethodHandle {
                reference_kind,
                reference_index,
            })
        }
        16 => {
            let descriptor_index = input.read_u16::<BigEndian>()?;

            Constant::MethodType(MethodType { descriptor_index })
        }
        17 => {
            let bootstrap_method_attr_index = input.read_u16::<BigEndian>()?;
            let name_and_type_index = input.read_u16::<BigEndian>()?;

            Constant::Dynamic(Dynamic {
                bootstrap_method_attr_index,
                name_and_type_index,
            })
        }
        18 => {
            let bootstrap_method_attr_index = input.read_u16::<BigEndian>()?;
            let name_and_type_index = input.read_u16::<BigEndian>()?;

            Constant::InvokeDynamic(InvokeDynamic {
                bootstrap_method_attr_index,
                name_and_type_index,
            })
        }
        19 => {
            let name_index = input.read_u16::<BigEndian>()?;

            Constant::Module(Module { name_index })
        }
        20 => {
            let name_index = input.read_u16::<BigEndian>()?;

            Constant::Package(Package { name_index })
        }
        _ => {
            return Err(ParseError::MatchOutOfBoundUsize(
                "constant tag",
                vec!["1..=20"],
                tag as usize,
            ))
        }
    };

    Ok(constant)
}
