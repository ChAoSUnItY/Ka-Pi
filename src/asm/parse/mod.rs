use std::fs::File;
use std::io::Read;
use std::ops::RangeFrom;
use std::path::Path;

use nom::combinator::map;
use nom::error::ParseError;
use nom::number::complete::be_u16;
use nom::{IResult, InputIter, InputLength, Slice};

use crate::asm::node::access_flag::AccessFlag;
use crate::asm::node::class::Class;
use crate::asm::node::constant::ConstantPool;
use crate::error::{KapiError, KapiResult};

pub mod attribute;
pub mod class;
pub mod constant;
pub mod field;
pub mod method;

pub fn read_class<P: AsRef<Path>>(class_path: P) -> KapiResult<Class> {
    let class_path = class_path.as_ref();
    let mut file = match File::open(class_path) {
        Ok(file) => file,
        Err(err) => {
            return Err(KapiError::ClassParseError(format!(
                "Unable to open class file {}, reason: {}",
                class_path.display(),
                err
            )))
        }
    };
    let mut class_bytes = Vec::new();

    if let Err(err) = file.read_to_end(&mut class_bytes) {
        return Err(KapiError::ClassParseError(format!(
            "Unable to read class file {}, reason: {}",
            class_path.display(),
            err
        )));
    }

    to_class(&class_bytes)
}

pub fn to_class(class_bytes: &[u8]) -> KapiResult<Class> {
    match class::class(&class_bytes[..]) {
        Ok((remain, class)) => {
            if !remain.is_empty() {
                Err(KapiError::ClassParseError(format!("Unable to parse class bytes, reason: class is fully parsed but there are {} bytes left, {remain:?}", remain.len())))
            } else {
                Ok(class)
            }
        }
        Err(err) => Err(KapiError::ClassParseError(format!(
            "Unable parse class bytes, reason: {err}"
        ))),
    }
}

fn collect<I, LP, TP, L, T, E: ParseError<I>>(
    mut len_parser: LP,
    mut item_parser: TP,
) -> impl FnMut(I) -> IResult<I, (L, Vec<T>), E>
where
    I: Slice<RangeFrom<usize>> + InputIter<Item = u8> + InputLength,
    LP: FnMut(I) -> IResult<I, L, E>,
    TP: FnMut(I) -> IResult<I, T, E>,
    L: Into<u64> + Copy,
{
    move |input: I| {
        let (mut input, len) = len_parser(input)?;
        let length = len.into();
        let mut items = Vec::with_capacity(length as usize);

        for _ in 0..length {
            let (remain, item) = item_parser(input)?;

            items.push(item);
            input = remain;
        }

        Ok((input, (len, items)))
    }
}

fn collect_with_constant_pool<'parser, 'constant_pool: 'parser, I, LP, TP, L, T, E: ParseError<I>>(
    mut len_parser: LP,
    mut item_parser: TP,
    constant_pool: &'constant_pool ConstantPool,
) -> impl FnMut(I) -> IResult<I, (L, Vec<T>), E> + '_
where
    I: Slice<RangeFrom<usize>> + InputIter<Item = u8> + InputLength,
    LP: FnMut(I) -> IResult<I, L, E> + 'parser + 'constant_pool,
    TP: FnMut(I, &'constant_pool ConstantPool) -> IResult<I, T, E> + 'parser + 'constant_pool,
    L: Into<u64> + Copy,
{
    move |input: I| {
        let (mut input, len) = len_parser(input)?;
        let length = len.into();
        let mut items = Vec::with_capacity(length as usize);

        for _ in 0..length {
            let (remain, item) = item_parser(input, constant_pool)?;

            items.push(item);
            input = remain;
        }

        Ok((input, (len, items)))
    }
}

fn access_flag<F>(input: &[u8]) -> IResult<&[u8], Vec<F>>
where
    F: AccessFlag,
{
    map(be_u16, F::mask_access_flags)(input)
}
