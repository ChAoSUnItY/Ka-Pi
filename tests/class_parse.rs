use insta::assert_yaml_snapshot;
use ka_pi::parse::{to_class, ParsingOption};
use std::io::Cursor;

use ka_pi::parse::ParseResult;

#[test]
fn test_main() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/Main.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}

#[test]
fn test_enum() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/Enum.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}

#[test]
fn test_record() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/Record.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}

#[test]
fn test_visible_annotation() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/VisibleAnnotation.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}

#[test]
fn test_invisible_annotation() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/InvisibleAnnotation.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}

#[test]
fn test_annotation_target() -> ParseResult<()> {
    let mut cursor = Cursor::new(include_bytes!(
        "../compiled_source/out/production/compiled_source/AnnotationTarget.class"
    ));

    assert_yaml_snapshot!(to_class(
        &mut cursor,
        ParsingOption::default().parse_attribute()
    )?);

    Ok(())
}
