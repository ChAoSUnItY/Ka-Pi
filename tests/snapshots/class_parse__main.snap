---
source: tests/class_parse.rs
expression: "to_class(include_bytes!(\"../compiled_source/out/production/compiled_source/Main.class\"))?"
---
java_version: V17
constant_pool_count: 60
constant_pool:
  len: 60
  entries:
    1:
      MethodRef:
        class_index: 2
        name_and_type_index: 3
    2:
      Class:
        name_index: 4
    3:
      NameAndType:
        name_index: 5
        type_index: 6
    4:
      Utf8:
        length: 16
        bytes:
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 79
          - 98
          - 106
          - 101
          - 99
          - 116
    5:
      Utf8:
        length: 6
        bytes:
          - 60
          - 105
          - 110
          - 105
          - 116
          - 62
    6:
      Utf8:
        length: 3
        bytes:
          - 40
          - 41
          - 86
    7:
      FieldRef:
        class_index: 8
        name_and_type_index: 9
    8:
      Class:
        name_index: 10
    9:
      NameAndType:
        name_index: 11
        type_index: 12
    10:
      Utf8:
        length: 16
        bytes:
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 83
          - 121
          - 115
          - 116
          - 101
          - 109
    11:
      Utf8:
        length: 3
        bytes:
          - 111
          - 117
          - 116
    12:
      Utf8:
        length: 21
        bytes:
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 105
          - 111
          - 47
          - 80
          - 114
          - 105
          - 110
          - 116
          - 83
          - 116
          - 114
          - 101
          - 97
          - 109
          - 59
    13:
      String:
        string_index: 14
    14:
      Utf8:
        length: 12
        bytes:
          - 72
          - 101
          - 108
          - 108
          - 111
          - 32
          - 119
          - 111
          - 114
          - 108
          - 100
          - 33
    15:
      MethodRef:
        class_index: 16
        name_and_type_index: 17
    16:
      Class:
        name_index: 18
    17:
      NameAndType:
        name_index: 19
        type_index: 20
    18:
      Utf8:
        length: 19
        bytes:
          - 106
          - 97
          - 118
          - 97
          - 47
          - 105
          - 111
          - 47
          - 80
          - 114
          - 105
          - 110
          - 116
          - 83
          - 116
          - 114
          - 101
          - 97
          - 109
    19:
      Utf8:
        length: 7
        bytes:
          - 112
          - 114
          - 105
          - 110
          - 116
          - 108
          - 110
    20:
      Utf8:
        length: 21
        bytes:
          - 40
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 83
          - 116
          - 114
          - 105
          - 110
          - 103
          - 59
          - 41
          - 86
    21:
      Class:
        name_index: 22
    22:
      Utf8:
        length: 5
        bytes:
          - 67
          - 108
          - 97
          - 115
          - 115
    23:
      MethodRef:
        class_index: 21
        name_and_type_index: 3
    24:
      FieldRef:
        class_index: 25
        name_and_type_index: 26
    25:
      Class:
        name_index: 27
    26:
      NameAndType:
        name_index: 28
        type_index: 29
    27:
      Utf8:
        length: 4
        bytes:
          - 77
          - 97
          - 105
          - 110
    28:
      Utf8:
        length: 5
        bytes:
          - 99
          - 108
          - 97
          - 122
          - 122
    29:
      Utf8:
        length: 7
        bytes:
          - 76
          - 67
          - 108
          - 97
          - 115
          - 115
          - 59
    30:
      FieldRef:
        class_index: 31
        name_and_type_index: 32
    31:
      Class:
        name_index: 33
    32:
      NameAndType:
        name_index: 34
        type_index: 35
    33:
      Utf8:
        length: 4
        bytes:
          - 69
          - 110
          - 117
          - 109
    34:
      Utf8:
        length: 4
        bytes:
          - 75
          - 65
          - 80
          - 73
    35:
      Utf8:
        length: 6
        bytes:
          - 76
          - 69
          - 110
          - 117
          - 109
          - 59
    36:
      FieldRef:
        class_index: 25
        name_and_type_index: 37
    37:
      NameAndType:
        name_index: 38
        type_index: 35
    38:
      Utf8:
        length: 1
        bytes:
          - 101
    39:
      Class:
        name_index: 40
    40:
      Utf8:
        length: 6
        bytes:
          - 82
          - 101
          - 99
          - 111
          - 114
          - 100
    41:
      MethodRef:
        class_index: 39
        name_and_type_index: 42
    42:
      NameAndType:
        name_index: 5
        type_index: 43
    43:
      Utf8:
        length: 18
        bytes:
          - 40
          - 66
          - 83
          - 73
          - 70
          - 74
          - 68
          - 90
          - 76
          - 82
          - 101
          - 99
          - 111
          - 114
          - 100
          - 59
          - 41
          - 86
    44:
      FieldRef:
        class_index: 25
        name_and_type_index: 45
    45:
      NameAndType:
        name_index: 46
        type_index: 47
    46:
      Utf8:
        length: 1
        bytes:
          - 114
    47:
      Utf8:
        length: 8
        bytes:
          - 76
          - 82
          - 101
          - 99
          - 111
          - 114
          - 100
          - 59
    48:
      Utf8:
        length: 4
        bytes:
          - 67
          - 111
          - 100
          - 101
    49:
      Utf8:
        length: 15
        bytes:
          - 76
          - 105
          - 110
          - 101
          - 78
          - 117
          - 109
          - 98
          - 101
          - 114
          - 84
          - 97
          - 98
          - 108
          - 101
    50:
      Utf8:
        length: 18
        bytes:
          - 76
          - 111
          - 99
          - 97
          - 108
          - 86
          - 97
          - 114
          - 105
          - 97
          - 98
          - 108
          - 101
          - 84
          - 97
          - 98
          - 108
          - 101
    51:
      Utf8:
        length: 4
        bytes:
          - 116
          - 104
          - 105
          - 115
    52:
      Utf8:
        length: 6
        bytes:
          - 76
          - 77
          - 97
          - 105
          - 110
          - 59
    53:
      Utf8:
        length: 4
        bytes:
          - 109
          - 97
          - 105
          - 110
    54:
      Utf8:
        length: 22
        bytes:
          - 40
          - 91
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 83
          - 116
          - 114
          - 105
          - 110
          - 103
          - 59
          - 41
          - 86
    55:
      Utf8:
        length: 4
        bytes:
          - 97
          - 114
          - 103
          - 115
    56:
      Utf8:
        length: 19
        bytes:
          - 91
          - 76
          - 106
          - 97
          - 118
          - 97
          - 47
          - 108
          - 97
          - 110
          - 103
          - 47
          - 83
          - 116
          - 114
          - 105
          - 110
          - 103
          - 59
    57:
      Utf8:
        length: 8
        bytes:
          - 60
          - 99
          - 108
          - 105
          - 110
          - 105
          - 116
          - 62
    58:
      Utf8:
        length: 10
        bytes:
          - 83
          - 111
          - 117
          - 114
          - 99
          - 101
          - 70
          - 105
          - 108
          - 101
    59:
      Utf8:
        length: 9
        bytes:
          - 77
          - 97
          - 105
          - 110
          - 46
          - 106
          - 97
          - 118
          - 97
access_flags:
  - Public
  - Super
this_class: 25
super_class: 2
interfaces_count: 0
interfaces: []
fields_count: 3
fields:
  - access_flags:
      - Public
      - Static
    name_index: 28
    descriptor_index: 29
    attribute_infos_len: 0
    attribute_infos: []
  - access_flags:
      - Public
      - Static
    name_index: 38
    descriptor_index: 35
    attribute_infos_len: 0
    attribute_infos: []
  - access_flags:
      - Public
      - Static
    name_index: 46
    descriptor_index: 47
    attribute_infos_len: 0
    attribute_infos: []
methods_count: 3
methods:
  - access_flags:
      - Public
    name_index: 5
    descriptor_index: 6
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 48
        attribute_len: 47
        info:
          - 0
          - 1
          - 0
          - 1
          - 0
          - 0
          - 0
          - 5
          - 42
          - 183
          - 0
          - 1
          - 177
          - 0
          - 0
          - 0
          - 2
          - 0
          - 49
          - 0
          - 0
          - 0
          - 6
          - 0
          - 1
          - 0
          - 0
          - 0
          - 1
          - 0
          - 50
          - 0
          - 0
          - 0
          - 12
          - 0
          - 1
          - 0
          - 0
          - 0
          - 5
          - 0
          - 51
          - 0
          - 52
          - 0
          - 0
        attribute:
          Code:
            max_stack: 1
            max_locals: 1
            code_length: 5
            code:
              - 42
              - 183
              - 0
              - 1
              - 177
            instructions:
              - ALOAD_0
              - INVOKESPECIAL:
                  index: 1
              - RETURN
            exception_table_length: 0
            exception_table: []
            attributes_length: 2
            attributes:
              - attribute_name_index: 49
                attribute_len: 6
                info:
                  - 0
                  - 1
                  - 0
                  - 0
                  - 0
                  - 1
                attribute:
                  LineNumberTable:
                    line_number_table_length: 1
                    line_number_table:
                      - start_pc: 0
                        line_number: 1
              - attribute_name_index: 50
                attribute_len: 12
                info:
                  - 0
                  - 1
                  - 0
                  - 0
                  - 0
                  - 5
                  - 0
                  - 51
                  - 0
                  - 52
                  - 0
                  - 0
                attribute:
                  LocalVariableTable:
                    local_variable_table_length: 1
                    local_variable_table:
                      - start_pc: 0
                        length: 5
                        name_index: 51
                        descriptor_index: 52
                        index: 0
  - access_flags:
      - Public
      - Static
    name_index: 53
    descriptor_index: 54
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 48
        attribute_len: 55
        info:
          - 0
          - 2
          - 0
          - 1
          - 0
          - 0
          - 0
          - 9
          - 178
          - 0
          - 7
          - 18
          - 13
          - 182
          - 0
          - 15
          - 177
          - 0
          - 0
          - 0
          - 2
          - 0
          - 49
          - 0
          - 0
          - 0
          - 10
          - 0
          - 2
          - 0
          - 0
          - 0
          - 7
          - 0
          - 8
          - 0
          - 8
          - 0
          - 50
          - 0
          - 0
          - 0
          - 12
          - 0
          - 1
          - 0
          - 0
          - 0
          - 9
          - 0
          - 55
          - 0
          - 56
          - 0
          - 0
        attribute:
          Code:
            max_stack: 2
            max_locals: 1
            code_length: 9
            code:
              - 178
              - 0
              - 7
              - 18
              - 13
              - 182
              - 0
              - 15
              - 177
            instructions:
              - GETSTATIC:
                  index: 7
              - LDC:
                  index: 13
              - INVOKEVIRTUAL:
                  index: 15
              - RETURN
            exception_table_length: 0
            exception_table: []
            attributes_length: 2
            attributes:
              - attribute_name_index: 49
                attribute_len: 10
                info:
                  - 0
                  - 2
                  - 0
                  - 0
                  - 0
                  - 7
                  - 0
                  - 8
                  - 0
                  - 8
                attribute:
                  LineNumberTable:
                    line_number_table_length: 2
                    line_number_table:
                      - start_pc: 0
                        line_number: 7
                      - start_pc: 8
                        line_number: 8
              - attribute_name_index: 50
                attribute_len: 12
                info:
                  - 0
                  - 1
                  - 0
                  - 0
                  - 0
                  - 9
                  - 0
                  - 55
                  - 0
                  - 56
                  - 0
                  - 0
                attribute:
                  LocalVariableTable:
                    local_variable_table_length: 1
                    local_variable_table:
                      - start_pc: 0
                        length: 9
                        name_index: 55
                        descriptor_index: 56
                        index: 0
  - access_flags:
      - Static
    name_index: 57
    descriptor_index: 6
    attribute_infos_len: 1
    attribute_infos:
      - attribute_name_index: 48
        attribute_len: 67
        info:
          - 0
          - 12
          - 0
          - 0
          - 0
          - 0
          - 0
          - 35
          - 187
          - 0
          - 21
          - 89
          - 183
          - 0
          - 23
          - 179
          - 0
          - 24
          - 178
          - 0
          - 30
          - 179
          - 0
          - 36
          - 187
          - 0
          - 39
          - 89
          - 4
          - 4
          - 4
          - 12
          - 10
          - 15
          - 4
          - 1
          - 183
          - 0
          - 41
          - 179
          - 0
          - 44
          - 177
          - 0
          - 0
          - 0
          - 1
          - 0
          - 49
          - 0
          - 0
          - 0
          - 14
          - 0
          - 3
          - 0
          - 0
          - 0
          - 2
          - 0
          - 10
          - 0
          - 3
          - 0
          - 16
          - 0
          - 4
        attribute:
          Code:
            max_stack: 12
            max_locals: 0
            code_length: 35
            code:
              - 187
              - 0
              - 21
              - 89
              - 183
              - 0
              - 23
              - 179
              - 0
              - 24
              - 178
              - 0
              - 30
              - 179
              - 0
              - 36
              - 187
              - 0
              - 39
              - 89
              - 4
              - 4
              - 4
              - 12
              - 10
              - 15
              - 4
              - 1
              - 183
              - 0
              - 41
              - 179
              - 0
              - 44
              - 177
            instructions:
              - NEW:
                  index: 21
              - DUP
              - INVOKESPECIAL:
                  index: 23
              - PUTSTATIC:
                  index: 24
              - GETSTATIC:
                  index: 30
              - PUTSTATIC:
                  index: 36
              - NEW:
                  index: 39
              - DUP
              - ICONST_1
              - ICONST_1
              - ICONST_1
              - FCONST_1
              - LCONST_1
              - DCONST_1
              - ICONST_1
              - ACONST_NULL
              - INVOKESPECIAL:
                  index: 41
              - PUTSTATIC:
                  index: 44
              - RETURN
            exception_table_length: 0
            exception_table: []
            attributes_length: 1
            attributes:
              - attribute_name_index: 49
                attribute_len: 14
                info:
                  - 0
                  - 3
                  - 0
                  - 0
                  - 0
                  - 2
                  - 0
                  - 10
                  - 0
                  - 3
                  - 0
                  - 16
                  - 0
                  - 4
                attribute:
                  LineNumberTable:
                    line_number_table_length: 3
                    line_number_table:
                      - start_pc: 0
                        line_number: 2
                      - start_pc: 10
                        line_number: 3
                      - start_pc: 16
                        line_number: 4
attributes_count: 1
attributes:
  - attribute_name_index: 58
    attribute_len: 2
    info:
      - 0
      - 59
    attribute:
      SourceFile:
        source_file_index: 59

